uber/jinja.py

```diff
@@ -1,4 +1,106 @@
 from uber.common import *
+import traceback
+import weakref
+from jinja2.loaders import split_template_path
+from jinja2.utils import open_if_exists
+
+
+class MultiPathEnvironment(jinja2.Environment):
+
+    def _find_requesting_templates(self):
+        """
+        Returns the set of all files in the call stack ending in ".html".
+        """
+        stack = map(lambda s: s[0], traceback.extract_stack())
+        return set(filter(lambda s: s.endswith('.html'), stack))
+
+    def _load_template(self, name, globals):
+        """
+        Overridden to take into consideration the templates in the call stack.
+        """
+        if self.loader is None:
+            raise TypeError('no loader for this environment specified')
+
+        requesting_templates = self._find_requesting_templates()
+
+        cache_key = (weakref.ref(self.loader), name, ','.join(requesting_templates))
+        if self.cache is not None:
+            template = self.cache.get(cache_key)
+            if template is not None and (not self.auto_reload or
+                                         template.is_up_to_date):
+                return template
+        template = self.loader.load(self, name, globals, requesting_templates)
+        if self.cache is not None:
+            self.cache[cache_key] = template
+        return template
+
+
+class MultiPathLoader(jinja2.FileSystemLoader):
+
+    def get_source(self, environment, template, requesting_templates=set()):
+        """
+        Overridden to accept the requesting_templates parameter: a set of every
+        template that has already been loaded during the current request.
+        """
+        pieces = split_template_path(template)
+        for searchpath in self.searchpath:
+            filename = os.path.join(searchpath, *pieces)
+            if filename in requesting_templates:
+                # If the file is already in the call stack, ignore it
+                continue
+
+            f = open_if_exists(filename)
+            if f is None:
+                continue
+            try:
+                contents = f.read().decode(self.encoding)
+            finally:
+                f.close()
+
+            mtime = os.path.getmtime(filename)
+
+            def uptodate():
+                try:
+                    return os.path.getmtime(filename) == mtime
+                except OSError:
+                    return False
+            return contents, filename, uptodate
+        raise TemplateNotFound(template)
```

Comment:
Just to double-check, will this class still work with the `ignore missing` keyword for the include tag? It's not a feature we use much (or at all...?) but it's a very nice feature.

Reply:
This should work the same way as the default jinja code - for the most part. Almost all the code in our custom classes is lifted straight from the jinja source.